README.md

```diff
@@ -1114,6 +1114,68 @@ If any of the modules failed to deploy, then Terragrunt will not attempt to depl
 **Note**: Not all blocks are able to access outputs passed by `dependency` blocks. See the section on
 [Configuration parsing order](#configuration-parsing-order) in this README for more information.
 
+##### Unapplied dependency and default outputs
+
+Terragrunt will return an error indicating the dependency hasn't been applied yet if the terraform module managed by the
+terragrunt config referenced in a `dependency` block has not been applied yet. This is because you cannot actually fetch
+outputs out of an unapplied Terraform module, even if there are no resources being created in the module.
+
+This is most problematic when running commands that do not modify state (e.g `plan-all` and `validate-all`) on a
+completely new setup where no infrastructure has been deployed. You won't be able to `plan` or `validate` a module if
+you can't determine the `inputs`. If the module depends on the outputs of another module that hasn't been applied
+yet, you won't be able to compute the `inputs` unless the dependencies are all applied. However, in real life usage, you
+would want to run `validate-all` or `plan-all` on a completely new set of infrastructure.
+
+To address this, you can provide default outputs to use when a module hasn't been applied yet. This is configured using
+the `default_outputs` attribute on the `dependency` block and it corresponds to a map that will be injected in place of
+the actual dependency outputs if the target config hasn't been applied yet.
+
+For example, in the previous example with a `mysql` module and `vpc` module, suppose you wanted to place in a temporary,
+dummy value for the `vpc_id` during a `validate-all` for the `mysql` module. You can specify in `mysql/terragrunt.hcl`:
+
+```
+dependency "vpc" {
+  config_path = "../vpc"
+
+  default_outputs = {
+    vpc_id = "temporary-dummy-id"
+  }
```

Comment:
Do `default_outputs` get used if the module has been applied, but doesn't have one of these outputs? Or _only_ if the module has never been applied at all?

Reply:
Also, will the user get a clear error if they forget to specify an output in `default_outputs`? E.g., let's say further down, they reference `depenency.vpc.outputs.public_subnet_ids`, but they never added `public_subnet_ids` to `default_outputs`. Will they immediately get an error that Terragrunt is using outputs, but one of the outputs they are looking up was not defined in `default_outputs`?

Reply:
> Do default_outputs get used if the module has been applied, but doesn't have one of these outputs? Or only if the module has never been applied at all?

Only if the module has never been applied at all. I think this is safer: having it pull from this default outputs on read and doesn't exist seem like it could cause some headaches.

> Also, will the user get a clear error if they forget to specify an output in default_outputs?

Ah that will not show a good error message, but this one is a bit harder to fix in the error because I can't pass through the info to the execution context, unless I implement what you suggest below.

> E.g., let's say further down, they reference depenency.vpc.outputs.public_subnet_ids, but they never added public_subnet_ids to default_outputs. Will they immediately get an error that Terragrunt is using outputs, but one of the outputs they are looking up was not defined in default_outputs?

Hmm I have to think a bit how this could be implemented. My guess is that it won't be trivial enough to include here in this PR, because I have to walk the tree to get the list of references and compare.

Reply:
NOTE: the following comment is no longer true

> Only if the module has never been applied at all. I think this is safer: having it pull from this default outputs on read and doesn't exist seem like it could cause some headaches.

Now, this will pull in the `default_outputs` if the module has not been applied, or if the module has no outputs.